crypto/bn/bn_exp2.c

```diff
@@ -32,7 +32,7 @@ int BN_mod_exp2_mont(BIGNUM *rr, const BIGNUM *a1, const BIGNUM *p1,
     bn_check_top(p2);
     bn_check_top(m);
 
-    if (!(m->d[0] & 1)) {
+    if (!BN_is_odd(m)) {
         BNerr(BN_F_BN_MOD_EXP2_MONT, BN_R_CALLED_WITH_EVEN_MODULUS);
```

Comment:
This error isn't quite accurate with this change. Just because it's not odd, doesn't make it even.

Reply:
The original bug was for where m is non-NULL but `m->d` is NULL.

It's plausible we should handle NULL m safely but we currently don't handle NULL a1, p1, a2, p2 or rr safely either.

Examining the BN code more generally, it doesn't seem like it's the done thing to error or assert on preconditions like "m is not NULL". In other words, it seems to be the convention in BN that unless otherwise specified, `BIGNUM*` arguments are mandatory and unchecked, with it being the callers responsibility to ensure the argument is non-NULL. Am I mistaken here? `bn_check_top` is commonly used but this only comes into play in debug builds.

Reply:
Yes, there is no need to check for m == NULL.

Reply:
>Just because it's not odd, doesn't make it even.

The `BN_is_odd()` is basically:
```
 return (a->top > 0) && (a->d[0] & 1);
 ```
If the value of `a->top == 0`, then it's not odd, and presumably not even either, since the `a->d` cannot be evaluated. The BN is in an invalid state. So, the error message should be something along the lines of "Called with invalid or even modulus", or just "Called with invalid modulus", which would include an even modulus.

It's really just a nit.

Reply:
IIRC `a->top == 0`, mean the BIGNUM is zero (not an invalid state)....I think??? Since 0 is even this gives the right result.

Reply:
Yes, @tmshort please look at BN_is_zero() implementation.

Reply:
Yeah, there's no explicit `BN_is_even()`. @t8m, so NVM